File: Applicationjava/mytablayout/src/main/java/com/example/mytablayout/MainActivity4.java
package com.example.mytablayout;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.ArrayList;

public class MainActivity4 extends AppCompatActivity {
    private ViewPager2 vp;
    private TabLayout tl;
    private TabLayoutMediator tabLayoutMediator;
    private ArrayList<String> mTitles = new ArrayList<>();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main4);

        vp = findViewById(R.id.vp);
        tl = findViewById(R.id.tl);
        tl.setTabMode(TabLayout.MODE_SCROLLABLE);

        mTitles.add("tab1");
        mTitles.add("tab2");
        mTitles.add("tab3");
        mTitles.add("tab4");
        mTitles.add("tab5");
        vp.setAdapter(new FragmentStateAdapter(this) {
            @NonNull
            @Override
            public Fragment createFragment(int position) {
                return new BlankFragment();
            }

            @Override
            public int getItemCount() {
                return mTitles.size();
            }

        });

        tabLayoutMediator = new TabLayoutMediator(tl, vp, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
//                tab.setText(mTitles.get(position));
                tab.setCustomView(getItemView(position));//自定义view
            }
        });
        tabLayoutMediator.attach();//绑定

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        tabLayoutMediator.detach();//解绑
    }

    public View getItemView(int postion){
        View v = LayoutInflater.from(this).inflate(R.layout.custom_tab_item,null);
        ImageView imageView = v.findViewById(R.id.imageView);
        TextView tv = v.findViewById(R.id.tv_title);
        tv.setText(mTitles.get(postion));
        return v;
    }
}

File: Applicationjava/mylifecycle/src/main/java/com/ly/mylifecycle/MainActivity4.kt
package com.ly.mylifecycle

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider

class MainActivity4 : AppCompatActivity() {
    val TAG = "MainActivity4"

    private lateinit var viewModel:MyViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main4)

        viewModel = ViewModelProvider(this).get(MyViewModel::class.java)
        Log.d(TAG, "onCreate: " + viewModel.currentName.value)

        val nameObserver = Observer<String>{
            Log.d(TAG, "onCreate: $it")
        }
        //这里传入了this作为lifeowner(activity实现了lifeownerde ), 进入函数里面可以知道 创建了LifecycleBoundObserver来观察当前activity即传入的owner，
        //LifecycleBoundObserver 实现了 LifecycleEventObserver 来监听状态变化 当activity状态变化会通知到这个实现方法并改变活跃状态等。。。当数据发生变化时候，判断活跃状态来进行是否要进行通知observer
        //所以在activity处于非活跃状态时候是不会收到数据变化通知的
        viewModel.currentName.observe(this,nameObserver)

        val mapObserver = Observer<String> {
            Log.d(TAG, "onCreate: $it")
        }
        viewModel.mapLiveData.observe(this,mapObserver)

    }

    fun btnClick(view: View) {
        viewModel.currentName.value = "hell world"
    }
}